src/code/ui/portal.js

```diff
@@ -0,0 +1,234 @@
+const WLPortal = L.Marker.extend({
+  type: "portal",
+
+  initialize: function (options) {
+    const operation = getSelectedOperation();
+    const portal = operation.getPortal(options.portalId);
+    options.title = portal.name;
+    L.Marker.prototype.initialize.call(this, portal.latLng, options);
+    this.bindPopup((layer) => layer._popupContent(), {
+      className: "wasabee-popup",
+      closeButton: false,
+    });
+
+    this.off("click", this._openPopup);
+    window.registerMarkerForOMS(this);
+    this.on("spiderfiedclick", this._openPopup);
+  },
+
+  _popupContent: function () {
+    const div = L.DomUtil.create("div", `wasabee-${this.type}-popup`);
+    return div;
+  },
+
+  _popupPortalComments: function (container, portal, canWrite) {
+    const portalComment = L.DomUtil.create(
+      "div",
+      "wasabee-portal-comment",
+      container
+    );
+    const pcLink = L.DomUtil.create("a", null, portalComment);
+    pcLink.textContent = portal.comment || wX("SET_PORTAL_COMMENT");
+    if (canWrite) {
+      pcLink.href = "#";
+      L.DomEvent.on(pcLink, "click", this._setPortalComment, this);
+    }
+    if (portal.hardness) {
+      const portalHardness = L.DomUtil.create(
+        "div",
+        "wasabee-portal-hardness",
+        container
+      );
+      const phLink = L.DomUtil.create("a", null, portalHardness);
+      phLink.textContent = portal.hardness;
+      if (canWrite) {
+        phLink.href = "#";
+        L.DomEvent.on(phLink, "click", this._setPortalComment, this);
+      }
+    }
+  },
+
+  _setPortalComment: function (ev) {
+    L.DomEvent.stop(ev);
+    const operation = getSelectedOperation();
+    const portal = operation.getPortal(this.options.portalId);
+    const scd = new SetCommentDialog({
+      target: portal,
+      operation: operation,
+    });
+    scd.enable();
+    this.closePopup();
+  },
+
+  _assignButton: function (container, text, target) {
+    const assignButton = L.DomUtil.create("button", null, container);
+    assignButton.textContent = text;
+    L.DomEvent.on(assignButton, "click", (ev) => {
+      L.DomEvent.stop(ev);
+      const ad = new AssignDialog({ target: target });
+      ad.enable();
+      this.closePopup();
+    });
+    return assignButton;
+  },
+
+  _deleteButton: function (container, text) {
+    const deleteButton = L.DomUtil.create("button", null, container);
+    deleteButton.textContent = text;
+    L.DomEvent.on(deleteButton, "click", (ev) => {
+      L.DomEvent.stop(ev);
+      if (this._deleteAction) this._deleteAction();
+      this.closePopup();
+    });
+  },
+
+  _sendTargetButton: function (container, text, target) {
+    const sendButton = L.DomUtil.create("button", null, container);
+    sendButton.textContent = text;
+    L.DomEvent.on(sendButton, "click", (ev) => {
+      L.DomEvent.stop(ev);
+      const std = new SendTargetDialog({ target: target });
+      std.enable();
+      this.closePopup();
+    });
+    return sendButton;
```

Comment:
why does _setTargetButton return the value but _deleteButton does not? if the caller doesn't consume the value, we should not return